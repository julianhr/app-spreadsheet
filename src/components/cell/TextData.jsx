import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'

import { clearCellValue } from '~/actions/tableActions'
import { setActiveCell } from '~/actions/globalActions'


const Text = styled.div`
  display: flex;
  align-items: center;
  outline: none;
  border: 2px solid transparent;
  width: 100%;
  height: 100%;
  padding: 0 2px;

  :focus {
    border: 2px solid salmon;
  }
`

export class TextData extends React.PureComponent {
  static propTypes = {
    // props
    isFocused: PropTypes.bool,
    location: PropTypes.string.isRequired,
    onDoubleClick: PropTypes.func.isRequired,
    onKeyDownEditable: PropTypes.func.isRequired,
    // redux
    clearCellValue: PropTypes.func.isRequired,
    setActiveCell: PropTypes.func.isRequired,
    text: PropTypes.string.isRequired,
  }

  constructor() {
    super()
    this.handleOnKeyDown = this.handleOnKeyDown.bind(this)
  }

  refText = React.createRef()

  componentDidMount() {
    this.focusTextTag()
  }

  focusTextTag() {
    if (this.props.isFocused) { this.refText.current.focus() }
  }

  handleOnKeyDown({ key }) {
    this.props.setActiveCell(this.props.location)

    if (['Delete', 'Backspace'].includes(key)) {
      if (this.props.text.length > 0) {
        this.props.clearCellValue(this.props.location)
      }
      // key pressed is a printable symbol, ex: 'a', '1', ','
      // can be further refined, but for now it's fine
    } else if (key.length === 1) {
      this.props.onKeyDownEditable()
    }
  }

  render() {
    return (
      <Text
        ref={this.refText}
        id={`t-${this.props.location}`}
        tabIndex='0'
        onKeyDown={this.handleOnKeyDown}
        onDoubleClick={this.props.onDoubleClick}
      >
        {this.props.text}
      </Text>
    )
  }
}

function mapStateToProps(state, ownProps) {
  const cell = state.table[ownProps.location]
  const text = cell ? cell.text : ''
  return { text }
}

const mapDispatchToProps = { clearCellValue, setActiveCell }

export default connect(mapStateToProps, mapDispatchToProps)(TextData)
