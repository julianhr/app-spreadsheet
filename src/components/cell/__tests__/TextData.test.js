import React from 'react'
import { create } from 'react-test-renderer'

import { TextData } from '../TextData'
import MockApp from '~/__tests__/__mocks__/MockApp'


const testProps = {
  // props
  isFocused: true,
  location: 'B-2',
  onDoubleClick: jest.fn(),
  onKeyDownEditable: jest.fn(),
  // redux
  clearCellValue: jest.fn(),
  setActiveCell: jest.fn(),
  text: 'test string',
}

const createApp = (props) => {
  jest.spyOn(TextData.prototype, 'focusTextTag').mockReturnValueOnce(null)
  const wrapper = create(<MockApp><TextData {...props} /></MockApp>)

  const cell = wrapper.root.findByType(TextData).instance
  cell.refText.current = document.createElement('div')
  return wrapper
}

describe('TextData', () => {
  describe('props', () => {
    test('all required props', () => {
      expect(() => createApp(testProps)).not.toThrow()
    })

    test('isFocused not required', () => {
      const props = {...testProps}
      delete props.isFocused
      expect(() => createApp(props)).not.toThrow()
    })

    test('location is required', () => {
      const props = {...testProps}
      delete props.location
      expect(() => createApp(props)).toThrow()
    })

    test('onDoubleClick is required', () => {
      const props = {...testProps}
      delete props.onDoubleClick
      expect(() => createApp(props)).toThrow()
    })

    test('onKeyDownEditable is required', () => {
      const props = {...testProps}
      delete props.onKeyDownEditable
      expect(() => createApp(props)).toThrow()
    })
  })

  describe('#focusTextTag', () => {
    it('if props.isFocused = true then focus text tag', () => {
      const props = {...testProps}
      props.isFocused = true
      const wrapper = createApp(props)
      const cell = wrapper.root.findByType(TextData).instance

      cell.refText.current = { focus: jest.fn() }
      cell.focusTextTag()
      expect(cell.refText.current.focus).toHaveBeenCalledTimes(1)
    })

    it('if props.isFocused = false then do not focus text tag', () => {
      const props = {...testProps}
      props.isFocused = false
      const wrapper = createApp(props)
      const cell = wrapper.root.findByType(TextData).instance

      cell.refText.current = { focus: jest.fn() }
      cell.focusTextTag()
      expect(cell.refText.current.focus).not.toHaveBeenCalled()
    })
  })

  describe('#handleOnKeyDown', () => {
    it('calls #setActiveCell', () => {
      const wrapper = createApp(testProps)
      const cell = wrapper.root.findByType(TextData).instance
      cell.handleOnKeyDown({ key: 'a' })
      expect(testProps.setActiveCell).toHaveBeenCalledTimes(1)
    })

    it('calls #clearCellValue if key pressed is Backspace or Delete', () => {
      const wrapper = createApp(testProps)
      const cell = wrapper.root.findByType(TextData).instance
      const keys = [
        ['Backspace', 1],
        ['Delete', 2],
        ['a', 2],
        ['1', 2],
        [',', 2]
      ]

      for (let [key, times] of keys) {
        cell.handleOnKeyDown({ key })
        expect(testProps.clearCellValue).toHaveBeenCalledTimes(times)
      }
    })

    it('calls #onKeyDownEditable if key pressed is printable symbol', () => {
      const wrapper = createApp(testProps)
      const cell = wrapper.root.findByType(TextData).instance
      const keys = [
        ['a', 1],
        ['1', 2],
        [',', 3],
        ['Backspace', 3],
        ['Delete', 3],
      ]

      for (let [key, times] of keys) {
        cell.handleOnKeyDown({ key })
        expect(testProps.onKeyDownEditable).toHaveBeenCalledTimes(times)
      }
    })
  })

  describe('functional tests', () => {
    it('calls #focusTextTag on mount', () => {
      const wrapper = createApp(testProps)
      const cell = wrapper.root.findByType(TextData).instance
      expect(cell.focusTextTag).toHaveBeenCalledTimes(1)
    })

    it('matches snapshot', () => {
      const wrapper = createApp(testProps)
      expect(wrapper).toMatchSnapshot()
    })
  })
})
