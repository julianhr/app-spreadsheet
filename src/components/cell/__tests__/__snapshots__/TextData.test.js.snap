// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextData functional tests matches snapshot 1`] = `
<div
  className="css-gchdh-Text e1anysog0"
  id="B-2"
  onDoubleClick={[MockFunction]}
  onKeyDown={[Function]}
  tabIndex="0"
>
  test string
</div>
`;
